File: playground/vm.py
frames = [(None, None)]
stack = []
bytecode = [
    # func 1 - calls add
    ('STACK_INIT', 0),
    ('PUSH_CONSTANT', 3),
    ('PUSH_CONSTANT', 2),
    ('CALL', 6),
    ('PRINT', None),
    ('RETURN', None),

    # func 2 - add
    ('STACK_INIT', 3),
    ('SET', 0),
    ('SET', 1),
    ('PUSH', 0),
    ('PUSH', 1),
    ('BIN_ADD', None),
    ('PRINT', None),
    ('SET', 2),
    ('PUSH', 2),
    ('PUSH_CONSTANT', 2),
    ('BIN_MUL', None),
    ('RETURN', None)

]

idx = 0
return_idx = -1

while idx < len(bytecode):
    if idx == -1:
        print('Program end')
        break

    if idx < 0:
        raise Exception('Invalid index {}'.format(idx))

    op, arg = bytecode[idx]
    method_stack = frames[-1][1]

    if op == 'PUSH_CONSTANT':
        stack.append(arg)
    elif op == 'SET':
        method_stack[arg] = stack.pop()
    elif op == 'PUSH':
        stack.append(method_stack[arg])
    elif op == 'BIN_ADD':
        stack.append(stack.pop() + stack.pop())
    elif op == 'BIN_MUL':
        stack.append(stack.pop() * stack.pop())
    elif op == 'RETURN':
        val = stack.pop()
        stack.append(val)
        idx = frames.pop()[0]
        continue
    elif op == 'CALL':
        assert bytecode[arg][0] == 'STACK_INIT'
        return_idx = idx + 1
        idx = arg
        continue

    elif op == 'STACK_INIT':
        frames.append((return_idx, [None] * arg))
    elif op == 'PRINT':
        print('Head of stack: {}'.format(stack[-1]))

    idx += 1

File: archives/vm.cpp
#include <cassert>
#include <stack>
#include <iostream>

enum OPCODES {
    STACK_INIT,
    PUSH_CONSTANT,
    PUSH,
    CALL,
    PRINT,
    RETURN,
    SET,
    BIN_ADD,
    BIN_MUL
};

struct Frame {
    int return_idx;
    int * data;
};

int main(){
    int program[][2] = {
            {STACK_INIT, 0},
            {PUSH_CONSTANT, 3},
            {PUSH_CONSTANT, 2},
            {CALL, 6},
            {PRINT, 0},
            {RETURN, 0},

            {STACK_INIT, 3},
            {SET, 0},
            {SET, 1},
            {PUSH, 0},
            {PUSH, 1},
            {BIN_ADD, 0},
            {PRINT, 0},
            {SET, 2},
            {PUSH, 2},
            {PUSH_CONSTANT, 2},
            {BIN_MUL, 0},
            {RETURN, 0}
    };

    int idx = 0;
    int return_idx = -1;
    int program_size = 100;


    std::stack<int> stack;
    std::stack<Frame> frames;
    Frame nf;
    frames.push(nf);

    while (idx != -1){
        assert(idx >= 0);

        OPCODES opcode = (OPCODES) program[idx][0];
        int arg = program[idx][1];

        int * frame = frames.top().data;
       // std::cout << "Index " << idx << std::endl;
        switch(opcode){
            case PUSH_CONSTANT:
                stack.push(arg);
                break;
            case SET:
                frame[arg] = stack.top();
                stack.pop();
                break;
            case PUSH:
                stack.push(frame[arg]);
                break;
            case BIN_ADD: {
                int lhs, rhs;
                lhs = stack.top();
                stack.pop();
                rhs = stack.top();
                stack.pop();
                stack.push(lhs + rhs);
                break;
            }
            case BIN_MUL: {
                int lhs, rhs;
                lhs = stack.top();
                stack.pop();
                rhs = stack.top();
                stack.pop();
                stack.push(lhs * rhs);
                break;
            }
            case RETURN:
                idx = frames.top().return_idx - 1;
                frames.pop();
                break;
            case CALL:
                return_idx = idx + 1;
                idx = arg - 1;
                break;
            case STACK_INIT: {
                Frame f = {return_idx, (int *) malloc(arg * sizeof(int *))};
                frames.push(f);
                break;
            }
            case PRINT:
                std::cout << "Head: " << stack.top() << std::endl;
                break;
        }

        idx += 1;



    }

}

